New Create2  Guide (#1070)
zeppelinos/zos: packages/docs/docs/docs/guide_create2_2.4.md

---
id: create2
title: Upgradable Create2 Contracts with zOS
---

> Create2 opcode gives us the ability to calculate smart contracts addresses without actually deploying them on Ethereum blockchain. This opens up a lot of possibilities to improve [user onboarding and scalability](https://blog.zeppelinos.org/getting-the-most-out-of-create2/).

Reviewer: ```suggestion
The CREATE2 opcode gives us the ability to calculate smart contracts addresses without actually deploying them on the Ethereum blockchain. This opens up a lot of possibilities to improve [user onboarding and scalability](https://blog.zeppelinos.org/getting-the-most-out-of-create2/).
```

We'll have a grammar editor go through the text later, but please keep an eye out for these kind of errors!

> In this guide, we will create an upgradableWallet contract with a transfer method, then we will reserve an address using Create2 without actually deploying Wallet contract and send some ethers to this address. At last, we will actually deploy our Wallet contract and execute transfer method to transfer contract funds to another account.

Reviewer: ```suggestion
In this guide, we will create an `upgradableWallet` contract with a `transfer` method, then we will reserve an address using Create2 without actually deploying Wallet contract and send some ethers to this address. At last, we will actually deploy our Wallet contract and execute transfer method to transfer contract funds to another account.
```

Let's use inline code where appropriate

Author: missed it...while converting it from Medium

## Ethereum Accounts

> Before diving into Create2, let’s understand a few basic things about Ethereum contract accounts. A contract account 

Reviewer: Re-reading this section, it doesn't seem to be relevant for anything that comes after - only that the CREATE opcode was used for creating a contract. I'd remove it altogether.

 zos init 
 ```

Next, we will create an upgradableWallet contract with a transfer method. We will reserve an address for it using Create2 and send some ethers to this contract without deploying it. In the end, we will actually deploy our Wallet contract and execute transfer method to transfer contract funds to another account.

## Add Wallet.sol

Next, we will add a Wallet.sol file in our contracts folder.
```solidity
pragma solidity ^ 0.5.0;
import "zos-lib/contracts/Initializable.sol";

contract Wallet is Initializable {
  address owner;
  function initialize(address _owner) initializer public {
    owner = _owner;
  }
  
  function transfer(address payable receiver) public {
    require(owner == msg.sender);
    receiver.transfer(address(this).balance);
  }
}
```

Now, we need to install and run [Ganache](https://docs.zeppelinos.org/docs/first.html#deploying-to-a-development-network) to deploy our contracts on local development blockchain.

## Deploying Contract using Create2

Before deploying, we need to add our contract usingzos add, this will add our contract to ZeppelinOS project, so it can be deployed using zos push afterward.

```console 
zos add Wallet 
```

> Now we will generate contract address using zos create2, we will use 12345 as salt, you can choose any random number and--query option tells zOS that we want to compute address, not the actual deployment of the contract. We will also pass network information using -n development , you can find this configuration innetworks.js file.

Reviewer: ```suggestion
Now we will generate contract address using zos create2, we will use 12345 as salt, you can choose any random number and --query option tells zOS that we want to compute address, not the actual deployment of the contract. We will also pass network information using -n development, you can find this configuration in networks.js file.
```
